crates/doc/src/shape/limits.rs

```diff
@@ -1,63 +1,180 @@
+                .object
+                .pattern_properties
+                .iter_mut()
+                .find(|pattern| regex_matches(&pattern.re, &prop_name))
+            {
+                pattern.shape = Shape::union(
+                    // Ideally we'd use a function like `replace_with` to allow replacing
+                    // pattern.shape with a value mapped from its previous value, but
+                    // that function doesn't exist yet. See https://github.com/rust-lang/rfcs/pull/1736
+                    // Instead, we must replace it with something temporarily while
+                    // Shape::union runs. Once it finishes, this `Shape::nothing()` is discarded.
+                    std::mem::replace(&mut pattern.shape, Shape::nothing()),
+                    shape_to_squash,
+                )
+            } else if let Some(addl_properties) = shape.object.additional_properties.take() {
+                shape.object.additional_properties =
+                    Some(Box::new(Shape::union(*addl_properties, shape_to_squash)));
+            } else {
+                shape.object.additional_properties = Some(Box::new(shape_to_squash))
+            }
         }
     }
 }
 
-pub const MAX_ROOT_FIELDS: usize = 750;
-pub const MAX_NESTED_FIELDS: usize = 200;
+fn squash_location(shape: &mut Shape, location: &[Token]) {
+    match location {
+        [] => unreachable!(),
+        [first] => squash_location_inner(shape, first),
+        [first, more @ ..] => {
+            let inner = resolve_shape_mut(shape, first.to_owned()).expect(&format!(
+                "Attempted to find property {first} that does not exist (more: {more:?})"
+            ));
+            squash_location(inner, more)
+        }
+    }
+}
+
+fn is_additionalx_field(token: &Token) -> bool {
+    match token {
+        Token::NextIndex => true,
+        Token::Property(prop_name) if prop_name == "*" => true,
+        _ => false,
+    }
+}
+
+/// Reduce the size/complexity of a shape while making sure that all
+/// objects that used to pass validation still do.
+pub fn enforce_shape_complexity_limit(shape: &mut Shape, limit: usize) {
+    let mut pointers = shape
+        .locations()
+        .into_iter()
+        .filter_map(|(ptr, _, _, _)| {
+            let last = ptr.0.last();
+            // Checking that the pointer has at least one item in it
+            if let Some(last) = last {
+                if !is_additionalx_field(last) {
+                    return Some(ptr);
+                }
+                return None;
+            } else {
+                None
+            }
+        })
+        .collect_vec();
+
+    if pointers.len() < limit {
+        return;
+    }
+
+    pointers.sort_by(|a_ptr, b_ptr| {
+        // order by depth, then lexicographically
+        match a_ptr.0.len().cmp(&b_ptr.0.len()) {
+            // Same depth, stably sort lexicographically
+            Ordering::Equal => a_ptr.to_string().cmp(&b_ptr.to_string()),
```

Comment:
optional overly-pedantic improvement: It'd make sense to me, to implement a total ordering for `doc::Pointer`.

There's still a small bit of weirdness where `/foo/9/stuff` will get pruned before `/foo/10/stuff`, which won't actually come up in practice because we don't infer tuples. It's also mildly more efficient (avoids string serialization).

Reply:
Ah, this was as simple as deriving `PartialOrd, Ord` for `doc::Pointer`, neat. Done!

Reply:
Oh, that's perfect. :heart: Rust.